src/components/motion/MotionController.cpp

```diff
@@ -29,31 +62,52 @@ void MotionController::Update(int16_t x, int16_t y, int16_t z, uint32_t nbSteps)
   this->nbSteps = nbSteps;
 }
 
-bool MotionController::ShouldRaiseWake(bool isSleeping) {
-  if ((x + 335) <= 670 && z < 0) {
-    if (!isSleeping) {
-      if (y <= 0) {
-        return false;
-      }
-      lastYForRaiseWake = 0;
-      return false;
-    }
+MotionController::AccelStats MotionController::GetAccelStats() const {
+  AccelStats stats;
 
-    if (y >= 0) {
-      lastYForRaiseWake = 0;
-      return false;
-    }
-    if (y + 230 < lastYForRaiseWake) {
-      lastYForRaiseWake = y;
-      return true;
-    }
+  for (uint8_t i = 0; i < AccelStats::numHistory; i++) {
+    stats.yMean += yHistory[histSize - i];
+    stats.zMean += zHistory[histSize - i];
+    stats.prevYMean += yHistory[1 + i];
+    stats.prevZMean += zHistory[1 + i];
   }
-  return false;
+  stats.yMean /= AccelStats::numHistory;
+  stats.zMean /= AccelStats::numHistory;
+  stats.prevYMean /= AccelStats::numHistory;
+  stats.prevZMean /= AccelStats::numHistory;
+
+  for (uint8_t i = 0; i < AccelStats::numHistory; i++) {
+    stats.yVariance += (yHistory[histSize - i] - stats.yMean) * (yHistory[histSize - i] - stats.yMean);
+    stats.zVariance += (zHistory[histSize - i] - stats.zMean) * (zHistory[histSize - i] - stats.zMean);
+  }
+  stats.yVariance /= AccelStats::numHistory;
+  stats.zVariance /= AccelStats::numHistory;
+
+  return stats;
+}
+
+bool MotionController::ShouldRaiseWake() const {
+  constexpr uint32_t varianceThresh = 56 * 56;
+  constexpr int16_t xThresh = 384;
+  constexpr int16_t yThresh = -64;
+  constexpr int16_t rollDegreesThresh = 45;
+
+  if (x < -xThresh || x > xThresh) {
+    return false;
+  }
+
+  // if the variance is below the threshold, the accelerometer values can be considered to be from acceleration due to gravity
+  if (stats.yVariance > varianceThresh || (stats.yMean < -724 && stats.zVariance > varianceThresh) || stats.yMean > yThresh) {
+    return false;
+  }
+
+  return DegreesRolled(stats.yMean, stats.zMean, stats.prevYMean, stats.prevZMean) > rollDegreesThresh;
 }
 
 bool MotionController::ShouldShakeWake(uint16_t thresh) {
   /* Currently Polling at 10hz, If this ever goes faster scalar and EMA might need adjusting */
-  int32_t speed = std::abs(z + (y / 2) + (x / 4) - lastY / 2 - lastZ) / (time - lastTime) * 100;
+  int32_t speed =
+    std::abs(zHistory[0] + (yHistory[0] / 2) + (x / 4) - yHistory[histSize - 1] / 2 - zHistory[histSize - 1]) / (time - lastTime) * 100;
```

Comment:
Maybe it's s good idea for documentation to split this into intermediate results with meaningful names.

Reply:
That's a good idea, but I think that that should be done in another PR. I also think that #1691 should be merged before that.